bitfield/src/lib.rs

```diff
@@ -10,6 +10,293 @@
+        type Out;
+    }
+
+    impl private::Sealed for [(); 0] {}
+    impl DispatchTrueFalse for [(); 0] {
+        type Out = False;
+    }
+
+    // impl private::Sealed for [(); 1] {} // <-- Already implemented by `define_specifiers` macro!
+    impl DispatchTrueFalse for [(); 1] {
+        type Out = True;
+    }
+
+    /// Public facing trait that is implemented by bitfield specifiers to
+    /// let the compiler check if all its variant discriminants are within
+    /// valid bounds.
+    pub trait CheckDiscriminantInRange
+    where
+        <Self::CheckType as DispatchTrueFalse>::Out: DiscriminantInRange
+    {
+        type CheckType: DispatchTrueFalse;
+    }
+
+    /// Helper type to check whether a bitfield member aligns to
+    /// the specified bits.
+    pub struct BitsCheck<A> {
+        pub arr: A,
+    }
+}
+
+/// Helper trait for underlying primitives handling of bitfields.
+///
+/// # Note
+///
+/// Must not and cannot be implemented by dependencies.
+pub trait PushBits: checks::private::Sealed {
+    fn push_bits(&mut self, amount: u32, bits: u8);
+}
+
+/// Helper trait for underlying primitives handling of bitfields.
+///
+/// # Note
+///
+/// Must not and cannot be implemented by dependencies.
+pub trait PopBits: checks::private::Sealed {
+    fn pop_bits(&mut self, amount: u32) -> u8;
+}
+
+macro_rules! impl_sealed_for {
+    ( $($primitive:ty),* ) => {
+        $(
+            impl checks::private::Sealed for $primitive {}
+        )*
+    }
+}
+
+impl_sealed_for!(bool, u8, u16, u32, u64, u128);
+
+impl PopBits for u8 {
+    #[inline(always)]
+    fn pop_bits(&mut self, amount: u32) -> u8 {
+        debug_assert!(amount <= 8);
+        let res = *self & ((0x1_u16.wrapping_shl(amount) as u8).wrapping_sub(1));
+        *self = self.wrapping_shr(amount);
+        res
+    }
+}
+
+macro_rules! impl_push_bits {
+    ( $($type:ty),+ ) => {
+        $(
+            impl PushBits for $type {
+                #[inline(always)]
+                fn push_bits(&mut self, amount: u32, bits: u8) {
+                    debug_assert!(amount <= 8);
+                    *self <<= amount;
+                    *self |= (bits & ((0x1_u16.wrapping_shl(amount) as u8).wrapping_sub(1))) as $type;
+                }
+            }
+        )+
+    }
+}
+
+impl_push_bits!(u8, u16, u32, u64, u128);
+
+macro_rules! impl_pop_bits {
+    ( $($type:ty),+ ) => {
+        $(
+            impl PopBits for $type {
+                #[inline(always)]
+                fn pop_bits(&mut self, amount: u32) -> u8 {
+                    debug_assert!(amount <= 8);
+                    let res = (*self & ((0x1 << amount) - 1)) as u8;
+                    *self >>= amount;
+                    res
+                }
+            }
+        )+
+    };
+}
+
+impl_pop_bits!(u16, u32, u64, u128);
+
+/// Trait implemented by primitives that drive bitfield manipulations generically.
+pub trait SpecifierBase: checks::private::Sealed {
+    type Base:
+        Default
```

Comment:
I don't think these trait bounds are used anywhere. This could be:

```rust
type Base;
```

Reply:
Unfortunately I actually use all of the three extensively in the expanded code.